cpp/velox/jni/JniWrapper.cc

```diff
@@ -154,6 +184,196 @@ JNIEXPORT void JNICALL Java_io_glutenproject_spark_sql_execution_datasources_vel
+    jboolean callBySelf) {
+  JNI_METHOD_START
+  auto splitter = shuffle_splitter_holder_.Lookup(splitter_id);
+  if (!splitter) {
+    std::string error_message = "Invalid splitter id " + std::to_string(splitter_id);
+    gluten::JniThrow(error_message);
+  }
+  jlong spilled_size;
+  gluten::JniAssertOkOrThrow(splitter->SpillFixedSize(size, &spilled_size), "(shuffle) nativeSpill: spill failed");
+  return spilled_size;
+  JNI_METHOD_END(-1L)
+}
+
+JNIEXPORT jlong JNICALL Java_io_glutenproject_vectorized_ShuffleSplitterJniWrapper_split(
+    JNIEnv* env,
+    jobject,
+    jlong splitter_id,
+    jint num_rows,
+    jlong handle) {
+  JNI_METHOD_START
+  auto splitter = shuffle_splitter_holder_.Lookup(splitter_id);
+  if (!splitter) {
+    std::string error_message = "Invalid splitter id " + std::to_string(splitter_id);
+    gluten::JniThrow(error_message);
+  }
+
+  // The column batch maybe VeloxColumnBatch or ArrowCStructColumnarBatch(FallbackRangeSplitter)
+  std::shared_ptr<ColumnarBatch> batch = ColumnBatchHolder::lookup(handle);
+  auto numBytes = batch->GetBytes();
+  gluten::JniAssertOkOrThrow(splitter->Split(batch), "Native split: splitter split failed");
+  return numBytes;
+  JNI_METHOD_END(-1L)
+}
+
+JNIEXPORT jobject JNICALL
+Java_io_glutenproject_vectorized_ShuffleSplitterJniWrapper_stop(JNIEnv* env, jobject, jlong splitter_id) {
+  JNI_METHOD_START
+  auto splitter = shuffle_splitter_holder_.Lookup(splitter_id);
+  if (!splitter) {
+    std::string error_message = "Invalid splitter id " + std::to_string(splitter_id);
+    gluten::JniThrow(error_message);
+  }
+
+  gluten::JniAssertOkOrThrow(splitter->Stop(), "Native split: splitter stop failed");
+
+  const auto& partition_lengths = splitter->PartitionLengths();
+  auto partition_length_arr = env->NewLongArray(partition_lengths.size());
+  auto src = reinterpret_cast<const jlong*>(partition_lengths.data());
+  env->SetLongArrayRegion(partition_length_arr, 0, partition_lengths.size(), src);
+
+  const auto& raw_partition_lengths = splitter->RawPartitionLengths();
+  auto raw_partition_length_arr = env->NewLongArray(raw_partition_lengths.size());
+  auto raw_src = reinterpret_cast<const jlong*>(raw_partition_lengths.data());
+  env->SetLongArrayRegion(raw_partition_length_arr, 0, raw_partition_lengths.size(), raw_src);
+
+  jobject split_result = env->NewObject(
+      split_result_class,
+      split_result_constructor,
+      0L,
+      splitter->TotalWriteTime(),
+      splitter->TotalSpillTime(),
+      splitter->TotalCompressTime(),
+      splitter->TotalBytesWritten(),
+      splitter->TotalBytesSpilled(),
+      partition_length_arr,
+      raw_partition_length_arr);
+
+  return split_result;
+  JNI_METHOD_END(nullptr)
+}
+
+JNIEXPORT void JNICALL
+Java_io_glutenproject_vectorized_ShuffleSplitterJniWrapper_close(JNIEnv* env, jobject, jlong splitter_id) {
+  JNI_METHOD_START
+  shuffle_splitter_holder_.Erase(splitter_id);
+  JNI_METHOD_END()
+}
+
 #ifdef __cplusplus
```

Comment:
Can we enrich the API `gluten::Backend` for shuffle functions? So that we can avoid moving the JNI wrapper across modules.

More about the cpp folders:

`cpp/core`: This folder is bound to Java folder `gluten-data` for placing common components for backends compatible to Arrow and Substrait

`cpp/velox`: This folder is bound to Java folder `backends-velox`

Reply:
Yes, have changed to this.